### PR TITLE
Motor Element

### PR DESCRIPTION
**Pull request by lthurner**
After discussion in #846, I revisited the introduction of the motor element as discussed in #244.

The motor element is integrated for power flow and short-circuit current.

@chefPony: I implented the motor short-circuit impedances as specified in the IEC 60909 standard

grafik

The  previous implementation of the impedance calculation was not correct. This version is now been successfully validated against PowerFactory.

### CODE REVIEW

**Comment by lthurner** [comment]
@chefPony @miek770 : can you please check and review?

File: pandapower/test/loadflow/test_scenarios.py [hunk 1]
```diff
@@ -507,5 +507,33 @@ def test_transformer_with_two_open_switches():
...
+    assert p == p_mech / efficiency * 100
+    assert p/s == cos_phi
+    res_bus_motor = net.res_bus.copy()
+
+    pp.create_load(net, b2, p_mw=net.res_motor.p_mw.values[0],
+                   q_mvar=net.res_motor.q_mvar.values[0])
+    net.motor.in_service = False
+
+    pp.runpp(net)
+    assert net.res_bus.equals(res_bus_motor)
```
Comment (chefPony) [inline pandapower/test/loadflow/test_scenarios.py:535]: Isn't this test a bit weak? It is not clear to me how it could ever fail. Can you compare with PowerFactory also here?

Comment (lthurner) [reply pandapower/test/loadflow/test_scenarios.py:535]: This just tests that all power values are internally added in the correct way with the correct sign to the power demand vector and results are correctly written in the res_motor dataframe (which is not automatically given).

You are correct that it does not test the motor model itself. I tried to validate against PowerFactory, but they have a much more complex model that works with slip curves, rotor impedances etc., so  was not able to reproduce this more simple model. Do you have an idea for a software to validate against?

File: pandapower/create.py [hunk 2]
```diff
@@ -1844,6 +1859,80 @@ def create_gens(net, buses, p_mw, vm_pu=1., sn_mva=nan, name=None, index=None, m
...
+        **pn_mech_mw** (float) - Mechanical rated power of the motor
+
+
+    OPTIONAL:
+
+        **name** (string, None) - The name for this motor
+
+        **cos_phi** (float, nan) - cosine phi at current operating point
+
+        **cos_phi_n** (float) - cosine phi at rated power of the motor for short-circuit calculation
```
[the remark this answers is not in the document]
Comment (lthurner) [reply pandapower/create.py:1884]: I think I made it optional because it isn't needed for power flow, but not for short-circuit calculation. But this is not in line with the other create functions, where all power flow parameters are always required. I will make it required.

**Comment by lthurner** [comment]
>     * can you enhance tests in `pandapower/test/shortcircuit/test_motor.py`? It would be great if you could add: a test for `case='min'` , and also for branch currents. For the branch currents I suggest to use a slightly more complicated topology, of the sort:
> 
> 
> ```
> ext_grid-o--line--o--line--o-load
>                   |
>                 motor
> ```

Yes will do as soon as possible, thanks for reviewing

**Comment by lthurner** [comment]
@chefPony: I added tests for minimum short-circuit current (motors are ignored) and branch results. I was not able to get external validation for the branch results, since PowerFactory does not provide the max/min branch results directly. I validated the branch results by checking consistency with the bus results. Can you check if you agree with the results?

**Comment by chefPony** [comment]
> Can you check if you agree with the results?

For case='min' I think it's fine since the motor are neglected according to IEC, for case='max' I don't have other software to compare with. So I tried to solve the circuit analytically and I get results that agree with yours. I report the solution below for future reference:

```
def motor_net():
    #      b1    b2     b3
    # eg---o-----o------o
    #            |
    #            M  
    net = pp.create_empty_network()
    b1 = pp.create_bus(net, vn_kv=0.4)
    b2 = pp.create_bus(net, vn_kv=0.4)
    b3 = pp.create_bus(net, vn_kv=0.4)
    pp.create_ext_grid(net, b1, s_sc_max_mva=10., rx_max=0.1, s_sc_min_mva=8.,
                       rx_min=0.1)
    pp.create_line_from_parameters(net, from_bus=b1, to_bus=b2, length_km=1.,
                                   r_ohm_per_km=0.32, c_nf_per_km=0,
                                   x_ohm_per_km=0.07, max_i_ka=1,
                                   endtemp_degree=80)
    
    pp.create_motor(net, b2, pn_mech_mw=0.5, lrc_pu=7., vn_kv=0.45, rx=0.4,
                    efficiency_n_percent=95, cos_phi_n=0.9, cos_phi=0.9)
    
    pp.create_line_from_parameters(net, from_bus=b2, to_bus=b3, length_km=2.,
                                   r_ohm_per_km=0.32, c_nf_per_km=0,
                                   x_ohm_per_km=0.07, max_i_ka=1,
                                   endtemp_degree=80)
    return net

# Constants
c = 1.1
net = motor_net()
Ibase = net.sn_mva / (net.bus.vn_kv.values * np.sqrt(3))
Zbase = (net.bus.vn_kv.values**2) / net.sn_mva

# ext_grid impedance
z_ext_grid_abs = c / net.ext_grid.s_sc_max_mva[0] 
x_ext_grid = z_ext_grid_abs / np.sqrt(1 + net.ext_grid.rx_max[0]**2) 
r_ext_grid = x_ext_grid * net.ext_grid.rx_max[0] 
z_ext_grid = (r_ext_grid + x_ext_grid*1j) * net.sn_mva

# line impedances
z_line1 =(net.line.r_ohm_per_km[0] + net.line.x_ohm_per_km[0]*1.j) * net.line.length_km[0] / Zbase[1]
z_line2 =(net.line.r_ohm_per_km[1] + net.line.x_ohm_per_km[1]*1.j) * net.line.length_km[1] / Zbase[2]

#motor impedance
sn_motor = net.motor.pn_mech_mw[0] / (net.motor.cos_phi_n[0] * net.motor.efficiency_n_percent[0] * 1e-2)

z_motor_abs = 1./net.motor.lrc_pu[0] * (net.motor.vn_kv[0]**2) / sn_motor
x_motor = z_motor_abs/np.sqrt(1 + net.motor.rx[0]**2)
r_motor = net.motor.rx[0] * x_motor
z_motor = (r_motor + x_motor*1j) / Zbase[1]

#bus currents
i_b1 = c / (z_motor + z_line1) + c / z_ext_grid
i_b1 = np.abs(i_b1 * Ibase[0])

i_b2 = c / (z_ext_grid + z_line1) + c / z_motor
i_b2 = np.abs(i_b2 * Ibase[1])

i_b3 = c / (z_line2 + z_motor * (z_ext_grid + z_line1) / (z_motor + z_ext_grid + z_line1))
i_b3 = np.abs(i_b3 * Ibase[2])

i_bus = np.array([i_b1, i_b2, i_b3])

# array([14.74380857,  5.62699407,  0.37073057])

# Branch currents
# Line 0
i_line_00 = np.abs(c / (z_motor + z_line1)) * Ibase[0]
i_line_10 = 0.
# Line 1
i_line_01 = np.abs(c / (z_ext_grid + z_line1)) * Ibase[1]
i_line_11 = 0.
#Line 2
i_line_12 = np.abs(c/((z_ext_grid + z_line1) * z_motor/(z_ext_grid + z_line1 + z_motor) + z_line2))
i_line_12 = np.abs(i_line_12) * Ibase[2]
i_line_02 = np.abs(i_line_12 * z_motor/(z_ext_grid + z_line1 + z_motor))
i_line_02 = np.abs(i_line_02) * Ibase[2]


i_line_mat = np.array([[i_line_00, i_line_01, i_line_02],
                       [i_line_10, i_line_11, i_line_12]]) 


#array([[0.7102266 , 0.76185819, 0.05019449],
#       [0.        , 0.        , 0.37073057]])

#expected_line_res_sc = [0.76185819, 0.37073057]

net = motor_net()
sc.calc_sc(net, case='max', branch_results=True)
ikss_with_motor = net.res_bus_sc.ikss_ka.copy()

net.motor.in_service = False
sc.calc_sc(net, case="max")
ikss_without_motor = net.res_bus_sc.ikss_ka.copy()
```

File: pandapower/test/shortcircuit/test_motor.py [hunk 3]
```diff
@@ -15,28 +15,61 @@ def motor_net():
...
 
-    assert np.isclose(net.res_bus_sc.ikss_ka.at[0], 14.743809197, rtol=1e-4)
-    assert np.isclose(net.res_bus_sc.ikss_ka.at[1], 5.626994278, rtol=1e-4)
+    # The maximum current through the first branch is the short-circuit current
+    # at the second bus without the motor contribution, which does not flow
+    # through the line
+    assert np.isclose(net.res_line_sc.ikss_ka.at[0], ikss_without_motor.at[1])
+    # The maximum current through the second branch is the short-circuit current
+    # at the third bus
+    assert np.isclose(net.res_line_sc.ikss_ka.at[1], net.res_bus_sc.ikss_ka.at[2])
```
Comment (chefPony) [inline pandapower/test/shortcircuit/test_motor.py:62]: It is probably better to put expected results as pure floats in the assert statements to limit test dependency on functions different from the tested one.

Comment (lthurner) [reply pandapower/test/shortcircuit/test_motor.py:64]: I included the bus results as floats in the test. I still want to keep the branch results checked against the bus results as a sanity check for internal consistency.

**Comment by lthurner** [comment]
> I think the only corner case left to test is when the network current contribution is lower than the motor one

Is that a specific corner case? I did include a test with a large motor now, but I didn't notice any structurally different behaviour.
